src/GafferScene/Constraint.cpp

```diff
@@ -38,11 +38,594 @@
+						"Gaffer::Constraint : Target primitive of type \"%s\" is not supported in Vertex target mode."
+					) % ( object.typeName() ) ).str() );
+			}
+			break;
+	}
+}
+
+void computeMeshUVLocalFrame( const IECoreScene::MeshPrimitive& primitive, Imath::M44f& m, const Imath::V2f& uv, const std::string& uvSet, const bool throwOnError, const IECore::Canceller* const canceller )
+{
+	const IECore::V3fVectorData* const pdata = primitive.variableData< IECore::V3fVectorData >( "P" );
+
+	if( pdata == 0 )
+	{
+		if( throwOnError )
+		{
+			throw IECore::InvalidArgumentException( "Gaffer::Contraint : MeshPrimitive has no Vertex \"P\" primitive variable." );
+		}
+		else
+		{
+			return;
+		}
+	}
+
+	const IECore::V3fVectorData::ValueType& points = pdata->readable();
+	const IECore::IntVectorData::ValueType& indices = primitive.vertexIds()->readable();
+	const IECore::IntVectorData::ValueType& faces = primitive.verticesPerFace()->readable();
+
+	const UVIndexer uvs( primitive, uvSet, throwOnError );
+	if( ! uvs.valid() )
+	{
+		return;
+	}
+
+	bool success = false;
+
+	for( int i = 0, nf = faces.size(), js = 0; i < nf; ++i )
+	{
+		const int ni = faces[ i ];
+
+		// canceller support
+
+		if( ( i % 100 ) == 0 )
+		{
+			IECore::Canceller::check( canceller );
+		}
+
+		// determine if uv coordinate is inside face
+
+		if( ptInPolygon( [ & uvs, js ]( const int ii ) { return uvs[ js + ii ]; }, uv, ni ) )
+		{
+			Imath::V3f fp( 0.f );
+
+			// triangulate face using subtracting ears method
+
+			IndexList findices;
+			std::unique_ptr< Index[] > values( new Index[ ni ] );
+			for( int j = 0; j < ni; ++j )
+			{
+				values[ j ].index = js + j;
+				findices.push_back( values[ j ] );
+			}
+
+			CircularIndexIterator it( & findices, ++( findices.begin() ) );
+			CircularIndexIterator itEnd( & findices, findices.begin() );
+
+			while( findices.size() > 2 )
+			{
+				bool clip = true;
+
+				if( it != itEnd )
+				{
+					CircularIndexIterator it2 = it;
+
+					const Imath::V2f tuv[ 3 ] =
+					{
+						uvs[ ( *( it2++ ) ).index ],
+						uvs[ ( *( it2++ ) ).index ],
+						uvs[ ( *( it2++ ) ).index ]
+					};
+
+					if( ( ( tuv[ 0 ] - tuv[ 1 ] ).cross( tuv[ 2 ] - tuv[ 1 ] ) >= 0.f ) )
+					{
+						clip = false;
+					}
+					else for( ; it2 != it; ++it2 )
+					{
+						if( ptInPolygon( [ & tuv ]( const int ii ){ return tuv[ ii ]; }, uvs[ ( *it2 ).index ], 3 ) )
+						{
+							clip = false;
+							break;
+						}
+					}
+				}
+
+				if( clip )
+				{
+					CircularIndexIterator tit = it;
+					itEnd = it;
+
+					const int i0 = ( *tit ).index; ++tit;
+					const int i1 = ( *tit ).index; tit = findices.erase( static_cast< IndexList::iterator >( tit ) );
+					const int i2 = ( *tit ).index;
+
+					const Imath::V2f tuv[ 3 ] =
+					{
+						uvs[ i0 ],
+						uvs[ i1 ],
+						uvs[ i2 ]
+					};
+
+					// test if uv in triangle
+
+					if( ptInPolygon( [ & tuv ]( const int ii ){ return tuv[ ii ]; }, uv, 3 ) )
+					{
+						success = true;
```

Comment:
John asked whether I thought this logic was too deeply nested ... I feel like I don't generally mind some deep nesting, but having the logic for what to do once we've found a triangle nested all the way down inside this fairly complex function for finding the right triangle does feel like it makes it a bit harder to wrap my head around.  It's also not clear why part of setting up the result is done outside this loop in the `if( success  )` branch.  Would it make sense to you if you got rid of the success flag, and instead did something like:

```
if( ptInPolygon( ... ) )
{
    m = computeTriangleUVLocalFrame( primitive, i0, i1, i2, tuv, uv );
    return;
}
```

with all the work once we find the triangle done in computeTriangleUVLocalFrame?  Actually, thinking about simple function signatures, is there a reason why all these *LocalFrame functions don't return M44f?  I tend to be cautious about copying around M44f's, but John usually tells me that I should be able to expect RVO these days.

Reply:
One other way I think it could be simplified, without even making new functions, would be to use `continue` :

```
if( !ptInPolygon( ... ) )
{
    // Maybe this increment can be moved to the `for()`,
    // so we don't have to remember to do it in two places?
    js += ni; 
    continue;
}

// Do triangulation
```

This would remove one level of nesting, but it also means a reader doesn't have to be considering whether or not there might be an `else` clause offscreen below.